source/common/network/transport_socket_options_impl.cc

```diff
@@ -40,46 +42,64 @@ void CommonUpstreamTransportSocketFactory::hashKey(
   for (const auto& protocol : alpn_fallback) {
     pushScalarToByteVector(StringUtil::CaseInsensitiveHash()(protocol), key);
   }
-}
 
-TransportSocketOptionsConstSharedPtr TransportSocketOptionsUtility::fromFilterState(
-    const StreamInfo::FilterStateSharedPtr& filter_state) {
-  if (!filter_state) {
-    return nullptr;
+  for (const auto& object : options->filterStateObjects()) {
+    if (auto hashable = dynamic_cast<const Hashable*>(object.data_.get()); hashable != nullptr) {
```

Comment:
What happens if an object isn't `Hashable`? Can we end up with bugs due to matching transport socket options hash even though it contains different filter state? Should `FilterState::Object` always inherit from `Hashable`? Or maybe make it required when the object opts in to sharing, via an accessor like `getHashable()`? We mostly try to avoid `dynamic_cast` in envoy for this type of feature detection.